Implementation of CCPP timestep_init and timestep_final phases

climbfuji opened the pull request:
This PR contains the following changes:

- move time vary physics from `run` to `timestep_init`
- remove dependency of time vary physics on GFS DDTs
- clean up o3 and h2o physics (use existing arrays instead of duplicating them and wasting memory)

Associated PRs:

https://github.com/ufs-community/ufs-weather-model/pull/337
https://github.com/NOAA-EMC/fv3atm/pull/217
https://github.com/NCAR/ccpp-physics/pull/534
https://github.com/NCAR/ccpp-framework/pull/344
https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere/pull/47

For regression testing information, see https://github.com/ufs-community/ufs-weather-model/pull/337.

climbfuji commented:
I am opening this PR for review. Although unlikely, I may have to pull in updates from NCAR master in case anything gets committed until it is the turn for this PR. But the code changes itself should be final.

tanyasmirnova reviewed (approved):
From all affected by this PR subroutines I am familiar with the gcycle.F90. Changes look good to me.

SMoorthi-emc commented:
Dom,
   I am not sure what changed and I don't see the updated code.
Besides, I am on leave until the beginning of next year.  So, may be I should not be reviewing this PR I guess.
Moorthi

climbfuji commented:
> Dom,
> I am not sure what changed and I don't see the updated code.
> Besides, I am on leave until the beginning of next year. So, may be I should not be reviewing this PR I guess.
> Moorthi

Absolutely not, I guess you shouldn't even be checking your emails ;-) No hurries at all, this is not yet a PR (just an issue noting the problem and suggesting a solution).

SMoorthi-emc reviewed (approved):
I have looked at the code, but cannot verify if everything is correct.
Nevertheless, I have to trust the CCPP team.

grantfirl commented:
@climbfuji Maybe I should know this, but for purposes of reviewing, is there any documentation that explains the new timestep_init and timestep_final CCPP phases? What is possible to do within them, what data should they have access to, etc.?

climbfuji commented:
> @climbfuji Maybe I should know this, but for purposes of reviewing, is there any documentation that explains the new timestep_init and timestep_final CCPP phases? What is possible to do within them, what data should they have access to, etc.?

I was looking at the ccpp-framework issues and meeting minutes, there is no good description of those. Maybe @gold2718 wrote a proposal back when we discussed it first time. In our meeting minutes, it appears first and with limited information on 06/06/2019, then again 06/04/2020:

https://github.com/NCAR/ccpp-framework/wiki/CCPP-Framework-Meeting-Minutes-2019-06-06
https://github.com/NCAR/ccpp-framework/wiki/CCPP-Framework-Meeting-Minutes-2020-06-04

In short: `timestep_init` and `timestep_final` captures processes that setup or finalize the physics for the current timestep. This is exactly what is in the current `time_vary` physics group. Similar as for the `init` and `finalize` phases, the rule is that no external loops over blocked data structures are allowed, i.e. all columns of an MPI rank must be passed at once. Threading is allowed inside the `timestep_init`/`timestep_final` physics, though.

The restriction is for the same reason as for the `init` and `final` phases, namely that such steps likely involve reading data (updated tables etc) from disk, or MPI synchronization.

climbfuji commented:
@SMoorthi-emc @grantfirl Thanks for reviewing. Note: So far, regression tests have passed on the following platforms against the existing baselines. This means that we can be fairly confident that the changes were made correctly. jet.intel is about to finish, same for cheyenne.intel. Denise is running the wcoss tests,
```
	modified:   tests/RegressionTests_cheyenne.gnu.log
	modified:   tests/RegressionTests_gaea.intel.log
	modified:   tests/RegressionTests_hera.gnu.log
	modified:   tests/RegressionTests_hera.intel.log
	modified:   tests/RegressionTests_orion.intel.log
```

climbfuji commented:
> @SMoorthi-emc @grantfirl Thanks for reviewing. Note: So far, regression tests have passed on the following platforms against the existing baselines. This means that we can be fairly confident that the changes were made correctly. jet.intel is about to finish, same for cheyenne.intel. Denise is running the wcoss tests,
> 
> ```
> 	modified:   tests/RegressionTests_cheyenne.gnu.log
> 	modified:   tests/RegressionTests_gaea.intel.log
> 	modified:   tests/RegressionTests_hera.gnu.log
> 	modified:   tests/RegressionTests_hera.intel.log
> 	modified:   tests/RegressionTests_orion.intel.log
> ```

Update: all regression tests passed on all tier-1 platforms. That's a good indication that we got the code changes right.

grantfirl reviewed (approved):
Approved

grantfirl commented on physics/GFS_rad_time_vary.fv3.F90:
```diff
@@ -6,99 +6,87 @@ module GFS_rad_time_vary
+             if (kdt == 1) then
+               t_2delt  = t
+               t_1delt  = t
+               qv_2delt = qv
```
This looks nice and clean. Since the RTs passed, I guess there was no need for the max function? Or, is it because there are no RTs using ZC MP anymore?

climbfuji replied:
No, we are running plenty of Zhao-Carr MP tests, including restart tests. They all passed. Means that the ICs already have a lower limit larger than that particular qmin (1.0E-10).

grantfirl replied:
Will that always be the case that ICs have already enforced a minimum value prior to this point in the physics? In other words, was the reason for removing it because it truly isn't necessary to enforce a minimum at this point? For a PR that is not supposed to change the answer, did we just get lucky?

climbfuji replied:
I think you are right. Let's fix this with your update of the corresponding SCM code. Thanks for finding this.

grantfirl commented on physics/GFS_phys_time_vary.fv3.F90:
```diff
@@ -32,247 +38,409 @@ module GFS_phys_time_vary
 
          if (is_initialized) return
 
-         nblks = size(Model%blksz)
```
I can't find an obvious place where this non-uniform block checking stuff would have gone. Is it magically handled in the CCPP framework now?

climbfuji replied:
No, this was a consistency check that was added a while back when the CCPP static mode was developed. It is no longer needed, because the logic in fv3atm takes care of it for the remainder of the lifetime of `ccpp_prebuild.py`.

grantfirl commented on physics/GFS_phys_time_vary.fv3.F90:
```diff
@@ -32,247 +38,409 @@ module GFS_phys_time_vary
+         if (nscyc >  0) then
+           if (mod(kdt,nscyc) == 1) THEN
+             call gcycle (me, nthrds, nx, ny, isc, jsc, nsst, tile_num, nlunit,       &
+                 input_nml_file, lsoil, lsoil_lsm, kice, idate, ialb, isot, ivegsrc,  &
```
This had to be a ton of fun passing all of these in!

grantfirl commented on physics/GFS_rrtmg_setup.F90:
```diff
@@ -523,12 +522,14 @@ subroutine radinit( si, NLAY, imp_physics, me )
 !> -# Set up control variables and external module variables in
 !!    module physparam
 #if 0
+      ! DH* WHAT IS THIS?
```
According to GitHub's blame function, this comment has existed since the original creation of the file, which was probably just copied from GFS_radiation_driver.F90 at the time. Maybe the comment was from Rusty Benson when he made IPD?

bensonr replied:
It most likely was something in from the radiation driver prior to the GFS physics cleanup that culminated in the IPD.

grantfirl commented on physics/GFS_suite_interstitial.F90:
```diff
@@ -16,15 +16,16 @@ end subroutine GFS_suite_interstitial_rad_reset_finalize
     subroutine GFS_suite_interstitial_rad_reset_run (Interstitial, Model, errmsg, errflg)
 
-      use GFS_typedefs, only: GFS_control_type,GFS_interstitial_type
+      use machine,      only: kind_phys
```
Why is kind_phys needed in this subroutine now?

climbfuji replied:
This is an oversight, it's not needed.

grantfirl commented:
I finally made my way through this -- really good work. I think that I understand it well enough to transfer the changes to SCM variants.

grantfirl commented on physics/GFS_phys_time_vary.fv3.F90:
```diff
@@ -316,215 +484,5 @@ subroutine GFS_phys_time_vary_finalize(errmsg, errflg)
-
-        !--- determine if diagnostics buckets need to be cleared
-        sec_zero = nint(Model%fhzero*con_hr)
-        if (sec_zero >= nint(max(Model%fhswr,Model%fhlwr))) then
```
It's funny that in the latest SCM PR where I added support for diagnostics intervals other than every timestep, I removed the clear calls from the host in favor of what was in the physics_time_vary run. Now, it will need to be added back. Not a big deal, just funny timing.
